File: apps/macos/Sources/SyncService.swift
import Foundation

protocol ClipboardReader {
    func readText() -> String?
}

protocol ClipboardWriter {
    func writeText(_ text: String)
}

protocol SecureStore {
    func get(_ key: String) -> String?
    func set(_ key: String, value: String)
    func delete(_ key: String)
}

final class SyncService {
    private let reader: ClipboardReader
    private let writer: ClipboardWriter
    private let store: SecureStore

    init(reader: ClipboardReader, writer: ClipboardWriter, store: SecureStore) {
        self.reader = reader
        self.writer = writer
        self.store = store
    }

    func captureClipboard() -> String? {
        return reader.readText()
    }

    func applyRemoteText(_ text: String) {
        writer.writeText(text)
    }

    func saveWorkspaceKey(_ key: String) {
        store.set("workspace_key", value: key)
    }

    func saveWebDavSettings(baseUrl: String, username: String, password: String, enabled: Bool) {
        store.set("webdav_enabled", value: enabled ? "1" : "0")
        store.set("webdav_base_url", value: baseUrl)
        store.set("webdav_username", value: username)
        store.set("webdav_password", value: password)
    }

    func loadWebDavSettings() -> (enabled: Bool, baseUrl: String, username: String, password: String) {
        (
            store.get("webdav_enabled") == "1",
            store.get("webdav_base_url") ?? "",
            store.get("webdav_username") ?? "",
            store.get("webdav_password") ?? ""
        )
    }

    func testWebDavConnection() async -> Bool {
        let cfg = loadWebDavSettings()
        guard cfg.enabled, let url = URL(string: cfg.baseUrl) else { return false }
        var request = URLRequest(url: url)
        request.httpMethod = "HEAD"
        applyBasicAuth(&request, username: cfg.username, password: cfg.password)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let http = response as? HTTPURLResponse {
                return (200...299).contains(http.statusCode)
            }
            return false
        } catch {
            return false
        }
    }

    func uploadClipboardToWebDav(_ text: String) async -> Bool {
        let cfg = loadWebDavSettings()
        guard cfg.enabled else { return false }
        let base = cfg.baseUrl.hasSuffix("/") ? cfg.baseUrl : cfg.baseUrl + "/"
        guard let url = URL(string: base + "clipboard-sync.txt") else { return false }

        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.httpBody = text.data(using: .utf8)
        request.setValue("text/plain; charset=utf-8", forHTTPHeaderField: "Content-Type")
        applyBasicAuth(&request, username: cfg.username, password: cfg.password)

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let http = response as? HTTPURLResponse {
                return (200...299).contains(http.statusCode)
            }
            return false
        } catch {
            return false
        }
    }

    func downloadClipboardFromWebDav() async -> String? {
        let cfg = loadWebDavSettings()
        guard cfg.enabled else { return nil }
        let base = cfg.baseUrl.hasSuffix("/") ? cfg.baseUrl : cfg.baseUrl + "/"
        guard let url = URL(string: base + "clipboard-sync.txt") else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        applyBasicAuth(&request, username: cfg.username, password: cfg.password)

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            if let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) {
                return String(data: data, encoding: .utf8)
            }
            return nil
        } catch {
            return nil
        }
    }

    private func applyBasicAuth(_ request: inout URLRequest, username: String, password: String) {
        guard !username.isEmpty else { return }
        let raw = Data("\(username):\(password)".utf8).base64EncodedString()
        request.setValue("Basic \(raw)", forHTTPHeaderField: "Authorization")
    }
}

File: apps/android/src/main/java/com/clipboardsync/SyncService.kt
package com.clipboardsync

import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.util.Base64

interface ClipboardReader {
    fun readText(): String?
}

interface ClipboardWriter {
    fun writeText(text: String)
}

interface SecureStore {
    fun get(key: String): String?
    fun set(key: String, value: String)
    fun delete(key: String)
}

class SyncService(
    private val reader: ClipboardReader,
    private val writer: ClipboardWriter,
    private val secureStore: SecureStore
) {
    fun captureClipboard(): String? = reader.readText()

    fun applyRemoteText(text: String) {
        writer.writeText(text)
    }

    fun saveWorkspaceKey(key: String) {
        secureStore.set("workspace_key", key)
    }

    fun saveWebDavSettings(baseUrl: String, username: String, password: String, enabled: Boolean) {
        secureStore.set("webdav_enabled", if (enabled) "1" else "0")
        secureStore.set("webdav_base_url", baseUrl)
        secureStore.set("webdav_username", username)
        secureStore.set("webdav_password", password)
    }

    fun loadWebDavSettings(): WebDavSettings {
        return WebDavSettings(
            enabled = secureStore.get("webdav_enabled") == "1",
            baseUrl = secureStore.get("webdav_base_url") ?: "",
            username = secureStore.get("webdav_username") ?: "",
            password = secureStore.get("webdav_password") ?: ""
        )
    }

    fun testWebDavConnection(): Boolean {
        val cfg = loadWebDavSettings()
        if (!cfg.enabled || cfg.baseUrl.isBlank()) return false
        return request(cfg.baseUrl, "HEAD", null, cfg) != null
    }

    fun uploadClipboardToWebDav(text: String): Boolean {
        val cfg = loadWebDavSettings()
        if (!cfg.enabled || cfg.baseUrl.isBlank()) return false
        val target = cfg.baseUrl.trimEnd('/') + "/clipboard-sync.txt"
        return request(target, "PUT", text, cfg) != null
    }

    fun downloadClipboardFromWebDav(): String? {
        val cfg = loadWebDavSettings()
        if (!cfg.enabled || cfg.baseUrl.isBlank()) return null
        val target = cfg.baseUrl.trimEnd('/') + "/clipboard-sync.txt"
        return request(target, "GET", null, cfg)
    }

    private fun request(url: String, method: String, body: String?, cfg: WebDavSettings): String? {
        return try {
            val conn = URL(url).openConnection() as HttpURLConnection
            conn.requestMethod = method
            conn.connectTimeout = 5000
            conn.readTimeout = 5000
            if (cfg.username.isNotBlank()) {
                val raw = Base64.getEncoder().encodeToString("${cfg.username}:${cfg.password}".toByteArray())
                conn.setRequestProperty("Authorization", "Basic $raw")
            }
            if (body != null) {
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "text/plain; charset=utf-8")
                conn.outputStream.use { it.write(body.toByteArray()) }
            }
            val code = conn.responseCode
            if (code in 200..299) {
                if (method == "GET") {
                    BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                } else {
                    "ok"
                }
            } else null
        } catch (_: Exception) {
            null
        }
    }
}

data class WebDavSettings(
    val enabled: Boolean,
    val baseUrl: String,
    val username: String,
    val password: String
)
